File: Level 3/다단계 칫솔 판매/solution.py
def solution(enroll, referral, sellers, amounts):
    parent = dict(zip(enroll, referral))

    total = {name: 0 for name in enroll}

    for seller, amount in zip(sellers, amounts):
        money = amount * 100

        while money > 0 and seller != "-":
            distribution_money = money // 10
            total[seller] += money - distribution_money
            seller = parent[seller]
            money = distribution_money

    return [total[name] for name in enroll]

File: Level 3/다단계 칫솔 판매/solution.js
function solution(enrolls, referrals, sellers, amounts) {
  const parent = getParent(enrolls, referrals);

  const total = getInitialTotal(enrolls);

  sellers.forEach((seller, index) => {
    let money = amounts[index] * 100;

    while (money > 0 && seller !== "-") {
      const distributionMoney = Math.floor(money / 10);
      total[seller] += money - distributionMoney;
      seller = parent[seller];
      money = distributionMoney;
    }
  });

  return enrolls.map((enroll) => total[enroll]);
}

function getParent(enrolls, referrals) {
  const parent = {};

  enrolls.forEach((enroll, index) => {
    parent[enroll] = referrals[index];
  });

  return parent;
}

function getInitialTotal(enrolls) {
  const total = {};

  enrolls.forEach((enroll) => {
    total[enroll] = 0;
  });

  return total;
}
